### PR TITLE
Fix Duplicate Scenario Runs

### LINKED ISSUE #40: Scenarios run multiple times
Tested with Cabbage 0.3.3 and Elixir 1.5.0.

When there are _n_ scenarios in a feature file, it seems that each scenario is run _n_ times. I've created a minimal test case [here](https://github.com/kerryb/cabbage_test), but basically what happens is this …

Starting with a single scenario, all seems fine:

```gherkin
Feature: Tests get run multiple times
  Scenario: Scenario 1
    Given a precondition
```

```
$ mix test

18:10:43.675 [debug] Parsing line: "Feature: Tests get run multiple times"

18:10:43.676 [debug] Parsing line: "Scenario: Scenario 1"

18:10:43.680 [debug] Parsing line: "Given a precondition"

18:10:43.682 [debug] Parsing line: ""

18:10:43.740 [info]  Line 2:  Scenario: Scenario 1

18:10:43.741 [info]  		Given a precondition
.

Finished in 0.1 seconds
1 test, 0 failures
```

Add a second scenario, and the number of tests jumps to four:

```gherkin
Feature: Tests get run multiple times
  Scenario: Scenario 1
    Given a precondition

  Scenario: Scenario 2
    Given a precondition
```

```
$ mix test

18:12:58.887 [debug] Parsing line: "Feature: Tests get run multiple times"

18:12:58.889 [debug] Parsing line: "Scenario: Scenario 1"

18:12:58.892 [debug] Parsing line: "Given a precondition"

18:12:58.894 [debug] Parsing line: ""

18:12:58.894 [debug] Parsing line: "Scenario: Scenario 2"

18:12:58.894 [debug] Parsing line: "Given a precondition"

18:12:58.894 [debug] Parsing line: ""

18:12:59.001 [info]  Line 2:  Scenario: Scenario 1

18:12:59.001 [info]  		Given a precondition
.
18:12:59.001 [info]  Line 2:  Scenario: Scenario 1

18:12:59.001 [info]  		Given a precondition
.
18:12:59.002 [info]  Line 5:  Scenario: Scenario 2

18:12:59.002 [info]  		Given a precondition
.
18:12:59.002 [info]  Line 5:  Scenario: Scenario 2

18:12:59.002 [info]  		Given a precondition
.

Finished in 0.1 seconds
4 tests, 0 failures
```

With three scenarios, the number jumps to nine, and so on.

I've added code to the step definition that writes to a file, to verify that the tests are actually running multiple times (rather than it being a problem with reporting). I've also verified that the behaviour is the same regardless of whether or not the scenarios share steps.

### PR DESCRIPTION
**Pull request by rawkode**
List comprehension was creating a cartesian product of both subjects. Going with primitive increment

Closes #40

### CODE REVIEW

File: lib/cabbage/feature.ex [hunk 1]
```diff
@@ -165,7 +165,11 @@ defmodule Cabbage.Feature do
     scenarios = Module.get_attribute(env.module, :scenarios) || []
     steps = Module.get_attribute(env.module, :steps) || []
     tags = Module.get_attribute(env.module, :tags) || []
-    for scenario <- scenarios, test_number <- 1..Kernel.length(scenarios) do
+    test_number = 0
+
+    for scenario <- scenarios do
+      test_number = test_number + 1
```
Comment (mgwidmann) [inline lib/cabbage/feature.ex]: does this work? I thought immutability put a damper in these kind of things... is this just rebinding over and over again?

Comment (mgwidmann) [reply lib/cabbage/feature.ex]: Doesnt seem like it works in the shell

```elixir
iex(2)> i = 0
0
iex(3)> for x <- 1..10 do
...(3)>   i = i + 1
...(3)>   IO.puts i
...(3)> end
1
1
1
1
1
1
1
1
1
1
[:ok, :ok, :ok, :ok, :ok, :ok, :ok, :ok, :ok, :ok]
```

This looks to be what you want:
```elixir
iex(4)> 100..110 |> Enum.with_index() |> Enum.each(fn {item, index} -> IO.puts index end)
0
1
2
3
4
5
6
7
8
9
10
:ok
```
So you should so something like: (untested)
```elixir
Enum.with_index(scenarios) |> Enum.each(fn {scenario, test_number} ->
   # all the rest of the stuff here
end)
```

**Comment by rawkode** [comment]
Ah, because the rebinding is within the `for` context, I guess?

Your syntax is much cleaner, I wasn't aware of `Enum.with_index`. I'll adapt.

Thanks for the tips, @mgwidmann

**Comment by rawkode** [comment]
Sorry for the delay, but back with you now, @mgwidmann

File: lib/cabbage/feature.ex [hunk 2]
```diff
@@ -165,9 +165,12 @@ defmodule Cabbage.Feature do
     scenarios = Module.get_attribute(env.module, :scenarios) || []
     steps = Module.get_attribute(env.module, :steps) || []
     tags = Module.get_attribute(env.module, :tags) || []
-    for scenario <- scenarios, test_number <- 1..Kernel.length(scenarios) do
+    test_number = 0
+
+
+    Enum.with_index(scenarios) |> Enum.each(fn {scenario, test_number} ->
```
Comment (mgwidmann) [inline lib/cabbage/feature.ex]: missing the closing `(` or delete this one... this doesn't compile

**Review by mgwidmann — CHANGES_REQUESTED** [review]

**Comment by rawkode** [comment]
@mgwidmann It would be great to get this merged, please :+1: 

Assuming I've not made a mess of it again :joy:

**Review by mgwidmann — APPROVED** [review]

**Comment by mgwidmann** [comment]
Sorry I missed this! Out of curiosity, why'd you flip the number and the scenario name?

**Comment by mgwidmann** [comment]
Released in `v0.3.4`!

**Comment by mgwidmann** [comment]
Sorry I've had to pull version `0.3.4` because it isnt actually running any tests. In the before compile, the `for` loop used to return the quoted code but `Enum.each` does not which means no code is actually injected at compile time. To fix it, it should be an `Enum.map`, but then all the tests fail with an error like:

```
** (UndefinedFunctionError) function Cabbage.DataTableTest."test 0. Data table represented as a list of maps test Data table represented as a list of maps (0) Data table represented as a list of maps"/1 is undefined or private. Did you mean one of:

           * "test Data table represented as a list of maps (0) Data table represented as a list of maps"/1
```

I can try to look into it tonight, but I don't have time at the moment.

**Comment by rawkode** [comment]
@mgwidmann How do we add a test that verifies the quote test is correct?

**Pull request merged**